src/parsec.rs

```diff
@@ -17,38 +17,110 @@ use crate::{
 use log::LogLevel;
 #[cfg(not(feature = "mock_parsec"))]
 use parsec as inner;
-use std::collections::{btree_map::Entry, BTreeMap};
+use std::{
+    collections::{btree_map::Entry, BTreeMap},
+    fmt, mem,
+};
 
 #[cfg(feature = "mock_parsec")]
 pub use crate::mock::parsec::{
-    init_mock, ConsensusMode, NetworkEvent, Observation, Proof, PublicId, SecretId,
+    init_mock, ConsensusMode, Error, NetworkEvent, Observation, Proof, PublicId, SecretId,
 };
 #[cfg(not(feature = "mock_parsec"))]
-pub use parsec::{ConsensusMode, NetworkEvent, Observation, Proof, PublicId, SecretId};
+pub use parsec::{ConsensusMode, Error, NetworkEvent, Observation, Proof, PublicId, SecretId};
 
 pub type Block = inner::Block<chain::NetworkEvent, id::PublicId>;
 pub type Parsec = inner::Parsec<chain::NetworkEvent, FullId>;
 pub type Request = inner::Request<chain::NetworkEvent, id::PublicId>;
 pub type Response = inner::Response<chain::NetworkEvent, id::PublicId>;
 
+// TODO: Needs to be determined from better data.
+#[cfg(not(feature = "mock_parsec"))]
+const PARSEC_SIZE_LIMIT: usize = 1_000_000_000;
+// Mock parsec request/responses are much smaller, so we need a lower limit.
+#[cfg(feature = "mock_parsec")]
+const PARSEC_SIZE_LIMIT: usize = 100;
+
+// Keep track of size in case we need to prune.
+#[derive(Default, Debug, PartialEq, Eq)]
+struct ParsecSizeCounter(usize);
+
+impl ParsecSizeCounter {
+    fn increase_size(&mut self, size: usize) {
+        self.0 += size;
+    }
+
+    fn needs_pruning(&self) -> bool {
+        self.0 > PARSEC_SIZE_LIMIT
+    }
+}
+
+impl fmt::Display for ParsecSizeCounter {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "size: {}", self.0)
+    }
+}
+
+struct ParsecAndCounter {
+    pub parsec: Parsec,
+    pub size_counter: ParsecSizeCounter,
+}
+
+impl ParsecAndCounter {
+    fn handle_request(
+        &mut self,
+        src: &id::PublicId,
+        req: Request,
+        log_ident: &LogIdent,
+    ) -> Result<Response, Error> {
+        self.count_size(mem::size_of_val(&req), log_ident);
+        self.parsec.handle_request(src, req)
+    }
+
+    fn handle_response(
+        &mut self,
+        src: &id::PublicId,
+        resp: Response,
+        log_ident: &LogIdent,
+    ) -> Result<(), Error> {
+        self.count_size(mem::size_of_val(&resp), log_ident);
+        self.parsec.handle_response(src, resp)
+    }
+
+    fn count_size(&mut self, size: usize, log_ident: &LogIdent) {
+        self.size_counter.increase_size(size);
+        trace!(
+            "{} - Parsec size is now estimated to: {}.",
+            log_ident,
```

Comment:
not sure whether it's worth to pass down this LogIdent via such couple of function interface.
we can still log the size info (if needed) on elder level to avoid pass it down?

Reply:
Do you mean move the `trace` to `Approved::handle_parsec_request/response`?

Reply:
well, I don't think that trace is mandatory. 
However, if you think such log is useful, then, yes.

Reply:
I prefer it like it is now though. Either remove the trace or keep it in the subfunction instead of duplicating the code and adding noise to the upper functions.

Reply:
I will vote for removing the trace then.

Reply:
I like having these `trace` statements, but I can remove them if people feel strongly about it